### PR TITLE
Prepare for py3 compatibility

### PR DESCRIPTION
**Pull request by boidolr**
I tried to to the minimal steps necessary to support python3.
What's still missing for this PR:
- [x] source level compatibility (unicode & config parser).

What's missing to get full compatibility with python3:
- [x] updating dependencies so they can be installed on python3.
- [x] updating to xapian >= 1.3.2.
- [x] testing of whoosh & xapian with python3.

@cbess: Do you have any preferences regarding how aggressive dependencies should be updated? The extremes would be just picking the latest version or trying to bump until it installs and starts.
Note that should the search engine be updated a reindex might be necessary.

This is _not_ finished yet and ready to merge! I'd like to have input before giving a complete compatible version.

### CODE REVIEW

File: core/sherlock/backends/whoosh_backend.py [hunk 1]
```diff
@@ -59,12 +59,11 @@ def index_file(self, filepath, *args, **kwargs):
         contents = safe_read_file(filepath)
         if contents is None:
             return
-        path = unicode(filepath)
```
Comment (boidolr) [inline core/sherlock/backends/whoosh_backend.py]: This function is not compatible with py3. As far as I tested, it removal seems not to change things:
- Unicode in file names was not supported before (threw errors in sqlite3 using python 2.7.10).
- Search (and retrieval) of terms containing unicode literals still works.

**Comment by cbess** [comment]
I usually use the latest version and work backwards, or review the change log and update it as needed to work with the latest. We could make a separate branch for *py3* compatibility and do the dependency upgrades. Thoughts?

**Comment by cbess** [comment]
thanks for initiating this upgrade. I really appreciate it.

**Comment by boidolr** [comment]
If possible I'd aim for one version which supports both py2 and py3 - maybe increase the minimum to 2.7 & 3.5 if necessary. Having said that a separate branch for the dependency bump might be nice as it would give the option to test that further before dropping it on everyone.

Right now the only problems still left to get the code here working across both mayor versions should be the unicode stuff & handling the configparser. If/when I get that too work I'd like to get those changes merged in master - which would be the aim of this PR.

Personally I think that adding py3 support is a good opportunity to go all out and take the latest (stable) dependencies which work across both py2 and py3 - so as you suggested I'd be taking the latest versions and testing whether they work.

I've updated the description of this PR a bit to clarify.

**Comment by boidolr** [comment]
Rebased against HEAD & hopefully got compatibility working. I am not so happy about the unicode import in [whoosh_backend.py#L27](https://github.com/boidolr/text-sherlock/blob/0fb6d7a02882c97089624d24baae110619902ccb/core/sherlock/backends/whoosh_backend.py#L27) though. Any suggestions?

**Comment by cbess** [comment]
Perhaps you can use `six.u` [docs](https://pythonhosted.org/six/#six.u)

**Comment by cbess** [comment]
As you may have saw, I updated it in 2e4695d

**Comment by boidolr** [comment]
That's what I first thought, too. Unfortunately that breaks python3 (because `six.u` fails to work for unicode strings).

**Comment by cbess** [comment]
Gotcha, when you say "break" what is the error?

**Comment by boidolr** [comment]
Uh, mixed it up, actually its py2 which has problems:
```
>>> six.u(u'str')

Traceback (most recent call last):
  File "<pyshell#2>", line 1, in <module>
    six.u(u'str')
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/six.py", line 468, in u
    return unicode(s, "unicode_escape")
TypeError: decoding Unicode is not supported
```

So it breaks some of the tests & searching with whoosh will fail, too.

**Comment by boidolr** [comment]
That could also mean that the call could be removed completely. In which cases was it necessary to have those unicode calls?

**Comment by cbess** [comment]
Thats a really good point, I can't recall any of them being required, so I think we can remove them.

**Comment by cbess** [comment]
I reviewed the changes and ran the tests, I think we fixed it. See d4897eb

Thoughts?

**Comment by boidolr** [comment]
Just tested these locally & getting an error when adding the path of a file to be indexed - so there was a reason to add the calls in the first place:
```
$ python --version
Python 2.7.10
$ python main.py --index=rebuild
...
Traceback (most recent call last):
  File "main.py", line 99, in <module>
    run()
  File "main.py", line 92, in run
    reindex()
  File "main.py", line 70, in reindex
    indexer.index_path(path)
  File "./core/sherlock/indexer.py", line 41, in index_path
    idxr.index_text(path)
  File "./core/sherlock/indexer.py", line 129, in index_text
    self.__index_path(path)
  File "./core/sherlock/indexer.py", line 135, in __index_path
    self.__index_dir(path)
  File "./core/sherlock/indexer.py", line 193, in __index_dir
    self.__index_file(path)
  File "./core/sherlock/indexer.py", line 203, in __index_file
    self._index.index_file(filepath, document_id=db_record.id)
  File "./core/sherlock/backends/whoosh_backend.py", line 69, in index_file
    self._writer.update_document(**doc)
  File ".virtualenvs/sherlock27/lib/python2.7/site-packages/whoosh/writing.py", line 356, in update_document
    self.add_document(**fields)
  File ".virtualenvs/sherlock27/lib/python2.7/site-packages/whoosh/filedb/filewriting.py", line 369, in add_document
    items = field.index(value)
  File ".virtualenvs/sherlock27/lib/python2.7/site-packages/whoosh/fields.py", line 218, in index
    raise ValueError("%r is not unicode or sequence" % value)
```
I've started out by cleaning my data subdirectory to force a rebuild. Are you experiencing these errors, too?

**Comment by cbess** [comment]
I started from a fresh install. I didn't see those errors. I'm also, because of the `virtualenv` running python `2.7.10`

**Comment by cbess** [comment]
I think its because the `yaml` entries may not be unicode

**Comment by cbess** [comment]
I think I fixed it, test out 7c8a119

**Comment by boidolr** [comment]
Going from [7c8a119](https://github.com/cbess/text-sherlock/commit/7c8a11976a68fc7c0176950e712175242e37459b):
The tests run fine for me, too - as before.
When starting and index operation though the same issue persists. I tested around - the unicode calls for both the path & base name of the file to be indexed when building the document (in `whoosh_backend.py`) seem to be needed on my end. Searching itself works without using them for terms containing both ascii and non ascii characters.
The failure happens for file names only consisting of ascii characters.

My dependencies:
```
$ pip list
CherryPy (3.2.2)
Flask (0.9)
flask-peewee (0.6.2)
Jinja2 (2.6)
path.py (9.0)
peewee (2.0.5)
pip (9.0.1)
Pygments (1.5)
PyYAML (3.12)
setuptools (30.4.0)
six (1.10.0)
Werkzeug (0.11.11)
wheel (0.29.0)
Whoosh (2.4.1)
wtf-peewee (0.2.6)
WTForms (2.1)
```

**Comment by cbess** [comment]
Just to make sure I understand you, the `python main.py --index=rebuild` now successfully execs on your end?

**Comment by boidolr** [comment]
No. main.py --test runs successfully, the rebuild does not.

C. Bess <notifications@github.com> schrieb am Sa., 10. Dez. 2016 21:09:

**Comment by cbess** [comment]
Interesting, the rebuild (with default settings) works fine on my end. What settings are you using?

**Comment by boidolr** [comment]
Also the default, with `debug=true`, `index_path` & `default_index_name` set.

**Comment by boidolr** [comment]
Anything else I should be trying? I am happy to test stuff, just a bit lost on how to proceed now.

**Comment by cbess** [comment]
Perhaps blow away your current directory (or check out a new build) and see if that fixes the issue, because it runs fine over here. I'll retry mine as well to double check.

**Comment by boidolr** [comment]
Well. New checkout, still the same problem as before.
It goes away when I readd the changes in [whoosh_backend.py](https://github.com/boidolr/text-sherlock/blob/0fb6d7a02882c97089624d24baae110619902ccb/core/sherlock/backends/whoosh_backend.py#L27). Did those make problems fore you initially?

**Comment by boidolr** [comment]
With this version & the changes from #35 and #36:
- tests run for python 2 and python 3
- indexing and searching works for whoosh and xapian in both python 2 and python 3

There are two places which needed special handling:
- path handling during indexing in whoosh as discussed previously. You were right, the problem is that the paths returned from the yaml config module are not in unicode by the module - so they have to be converted if necessary.
- value handling in `xapian_backend.py`: as the python3 bindings return `bytes` instead of `str` they have to be decoded in python 3. I have chosen to use UTF-8 as encoding as it is used in all places where encodings are explicitly used/ - this might be wrong though and could depend on the actual input files.

**Comment by cbess** [comment]
Awesome! Lots of changes. Thanks for handling this, I'm going to clone this and review it locally before I merge. Hopefully I can do it in the next few days.

**Comment by boidolr** [comment]
If you do not mind resetting locally I would rebase against the current
HEAD and squash a bit. The resulting diffs  should be cleaner and easier to
read.

C. Bess <notifications@github.com> schrieb am Do., 5. Jan. 2017 05:33:

**Comment by cbess** [comment]
I've reviewed to code. Thanks for the contributions.

**Comment by cbess** [comment]
I'll squash and merge. What name would like to see as a contributor?

**Comment by boidolr** [comment]
'Raphael Boidol' is fine, thank you.

**Pull request merged**